chia/full_node/mempool_check_conditions.py

```diff
@@ -55,17 +60,26 @@ def get_name_puzzle_conditions(
         return NPCResult(uint16(Err.GENERATOR_RUNTIME_ERROR.value), None, uint64(0))
 
 
-def get_puzzle_and_solution_for_coin(generator: BlockGenerator, coin_name: bytes, max_cost: int):
+def get_puzzle_and_solution_for_coin(
+    generator: BlockGenerator, coin: Coin
+) -> Tuple[Optional[Exception], Optional[Program], Optional[Program]]:
     try:
-        block_program = generator.program
-        block_program_args = create_generator_args(generator.generator_refs)
-
-        cost, result = GENERATOR_FOR_SINGLE_COIN_MOD.run_with_cost(
-            max_cost, block_program, block_program_args, coin_name
-        )
-        puzzle = result.first()
-        solution = result.rest().first()
-        return None, puzzle, solution
+        args = [bytes(a) for a in generator.generator_refs]
+        cost, result = generator.program.run_with_cost(DEFAULT_CONSTANTS.MAX_BLOCK_COST_CLVM, DESERIALIZE_MOD, args)
+
+        requested_parent = Program.to(coin.parent_coin_info)
+        requested_amount = Program.to(coin.amount)
+
+        coin_spends = result.first()
+        for spend in coin_spends.as_iter():
+            parent, puzzle, amount, solution = spend.as_iter()
```

Comment:
This may be over the top but is there a chance that `as_iter()` does some extra work that we wouldn't need if the spend we're looking for is one of the first ones?  I can't imagine `first()` and `rest()` are that expensive but we _could_ technically do fewer of them if we only do `rest()` after we have decided `first()` is not the spend we're looking for.  Anyways, probably not worth it from a code complexity standpoint.

Reply:
I'm hoping to replace this with a rust function where all of this state and searching it will stay on the rust side

Reply:
hm. removing it now does *not* make mypy complain. But I don't fully trust that mypy always find things like this